power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/dataset.hpp

```diff
@@ -0,0 +1,170 @@
+            std::string(PGM_dataset_info_component_name(info.handle_.get(), info.info_.get(), component_idx));
+        info.handle_.check_error();
+        return component_name;
+    }
+    std::string const component_name(Idx component_idx) const { return component_name(*this, component_idx); }
+
+    static Idx component_elements_per_scenario(Info const& info, Idx component_idx) {
+        auto const elements_per_scenario =
+            PGM_dataset_info_elements_per_scenario(info.handle_.get(), info.info_.get(), component_idx);
+        info.handle_.check_error();
+        return elements_per_scenario;
+    }
+    Idx component_elements_per_scenario(Idx component_idx) const {
+        return component_elements_per_scenario(*this, component_idx);
+    }
+
+    static Idx component_total_elements(Info const& info, Idx component_idx) {
+        auto const total_elements =
+            PGM_dataset_info_total_elements(info.handle_.get(), info.info_.get(), component_idx);
+        info.handle_.check_error();
+        return total_elements;
+    }
+    Idx component_total_elements(Idx component_idx) const { return component_total_elements(*this, component_idx); }
+
+  private:
+    Handle handle_{};
+    std::unique_ptr<PGM_DatasetInfo const> info_;
+};
+
+class DatasetConst {
+  public:
+    DatasetConst(std::string const& dataset, Idx is_batch, Idx batch_size)
+        : dataset_{PGM_create_dataset_const(handle_.get(), dataset.c_str(), is_batch, batch_size)} {}
+    DatasetConst(DatasetWritable const& writable_dataset)
+        : dataset_{PGM_create_dataset_const_from_writable(handle_.get(), writable_dataset.get())} {}
+    DatasetConst(DatasetMutable const& mutable_dataset)
+        : dataset_{PGM_create_dataset_const_from_mutable(handle_.get(), mutable_dataset.get())} {}
+
+    RawConstDataset* get() const { return dataset_.get(); }
+
+    static void add_buffer(DatasetConst& dataset, std::string const& component, Idx elements_per_scenario,
+                           Idx total_elements, Idx const* indptr, RawDataConstPtr data) {
+        PGM_dataset_const_add_buffer(dataset.handle_.get(), dataset.dataset_.get(), component.c_str(),
+                                     elements_per_scenario, total_elements, indptr, data);
+        dataset.handle_.check_error();
+    }
+    void add_buffer(std::string const& component, Idx elements_per_scenario, Idx total_elements, Idx const* indptr,
+                    RawDataConstPtr data) {
+        return add_buffer(*this, component, elements_per_scenario, total_elements, indptr, data);
+    }
+
+    static Info const get_info(DatasetConst const& dataset) {
+        auto const info = PGM_dataset_const_get_info(dataset.handle_.get(), dataset.dataset_.get());
+        dataset.handle_.check_error();
+        return Info(info);
+    }
+    Info const get_info() const { return get_info(*this); }
+
+  private:
+    Handle handle_{};
+    detail::UniquePtr<RawConstDataset, PGM_destroy_dataset_const> dataset_;
```

Comment:
Optional: maybe it looks nice to do something like `using DatasetPtr = detail::UniquePtr<RawConstDataset, PGM_destroy_dataset_const>;` here and similarly in other places.

Reply:
I think this would be nice if such pointers were used elsewhere, but since they are private and their definition is the only place where the explicit type is used (for now), I say to leave as is, as it does not really add any value. If this changes, the developer can add it.